src/new-storage/DeleteExpiredCmd.js

```diff
@@ -0,0 +1,139 @@
+            }
+        })
+
+        // used for limited concurrency
+        this.limit = pLimit(5)
+    }
+
+    async _getStreams() {
+        const result = []
+
+        const query = 'SELECT DISTINCT stream_id, partition FROM bucket'
+        const resultSet = await this.cassandraClient.execute(query)
+
+        if (resultSet.rows.length) {
+            resultSet.rows.forEach((row) => {
+                result.push({
+                    streamId: row.stream_id,
+                    partition: row.partition
+                })
+            })
+        }
+
+        return result
+    }
+
+    async _fetchStreamsInfo(streams) {
+        const tasks = streams.map((stream) => {
+            return this.limit(async () => {
+                const url = `${this.baseUrl}/api/v1/streams/${stream.streamId}/validation`
+                return fetch(url).then((res) => res.json()).then((json) => {
+                    return {
+                        streamId: stream.streamId,
+                        partition: stream.partition,
+                        storageDays: parseInt(json.storageDays)
+                    }
+                }).catch((e) => console.error(e))
+            })
+        })
+
+        return Promise.all(tasks)
+    }
+
+    async _deleteExpired(streamsInfo) {
+        const tasks = streamsInfo.map((stream) => {
+            const { bucketId, dateCreate, streamId, partition } = stream
+            const queries = [
+                {
+                    query: 'DELETE FROM bucket WHERE stream_id = ? AND partition = ? AND date_create = ?',
+                    params: [streamId, partition, dateCreate]
+                },
+                {
+                    query: 'DELETE FROM stream_data WHERE stream_id = ? AND partition = ? AND bucket_id = ?',
+                    params: [streamId, partition, bucketId]
+                }
+            ]
+
+            return this.limit(async () => {
+                await this.cassandraClient.batch(queries, {
+                    prepare: true
+                }).catch((err) => {
+                    console.error(err)
+                })
+            })
+        })
+
+        return Promise.all(tasks)
+    }
+
+    async _getExpiredBuckets(streamsInfo) {
+        const result = []
+
+        const query = 'SELECT * FROM bucket WHERE stream_id = ? AND partition = ? AND date_create <= ?'
+
+        const tasks = streamsInfo.map((stream) => {
+            const { streamId, partition, storageDays } = stream
+            const timestampBefore = Date.now() - 1000 * 60 * 60 * 24 * storageDays
+            const params = [streamId, partition, timestampBefore]
+
+            return this.limit(async () => {
+                const resultSet = await this.cassandraClient.execute(query, params, {
+                    prepare: true,
+                }).catch((e) => console.error(e))
+
+                resultSet.rows.forEach((row) => {
+                    result.push({
```

Comment:
Should `result` be sorted somehow? This fires off a bunch of async calls which fill `result` but this doesn't guarantee they will resolve in the same order they're sent in. Does that matter?

Suggestion: `Promise.all` returns results in the order they appear in the `Promise.all`, so perhaps collect the result arrays, let `Promise.all` handle the ordering, then call `.flat` on them (code below totally untested):

```js
async _getExpiredBuckets(streamsInfo) {
    const query = 'SELECT * FROM bucket WHERE stream_id = ? AND partition = ? AND date_create <= ?'

    const result = await Promise.all(streamsInfo.map((stream) => {
        const { streamId, partition, storageDays } = stream
        const timestampBefore = Date.now() - 1000 * 60 * 60 * 24 * storageDays
        const params = [streamId, partition, timestampBefore]

        return this.limit(async () => {
            const resultSet = await this.cassandraClient.execute(query, params, {
                prepare: true,
            }).catch((e) => console.error(e))

            return !(resultSet && resultSet.rows) ? [] : resultSet.rows.map((row) => ({
                bucketId: row.id,
                dateCreate: row.date_create,
                streamId: row.stream_id,
                partition: row.partition,
                storageDays
            }))
        })
    }))

    return result.flat()
}
```

Reply:
We don't care about order